### PR TITLE
Plugin system

### PR DESCRIPTION
**Pull request by asutoshpalai**
This is the working PR for plugin system for bundler.

A rough workflow shall be
- [x] Add functionality to install a plugin as gem from RubyGem source and also from git source
- [x] Add support for command line plugin
- [ ] Add support for source plugin
- [ ] Add support for life cycle hooks
- [ ]  Finalize all the functionalities

### CODE REVIEW

File: lib/bundler/cli/plugin.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+module Bundler
+  class CLI::Plugin < Thor
+    desc "install PLUGIN", "Install the plugin from the source"
+    long_desc <<-D
+      Install a plugin named PLUGIN wither from the rubygems source provided (with --source option) or from a git source provided with (--git option).
+    D
+    method_option "source", :type => :string, :default => nil, :banner =>
+      "URL of the RubyGems source to fetch the plugin from"
+    method_option "version", :type => :string, :default => nil, :banner =>
+      "The version of the plugin to fetch from"
```
Comment (segiddins) [inline lib/bundler/cli/plugin.rb]: ~~from~~

File: lib/bundler/dsl.rb [hunk 2]
```diff
@@ -19,7 +19,7 @@ def self.evaluate(gemfile, lockfile, unlock)
 
     def initialize
       @source               = nil
-      @sources              = SourceList.new
+      @sources            ||= SourceList.new
```
Comment (segiddins) [inline lib/bundler/dsl.rb]: why is this change necessary?

File: lib/bundler/plugin.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+module Bundler
+  class Plugin
+    autoload :Dsl,        "bundler/plugin/dsl"
+    autoload :Index,      "bundler/plugin/index"
+    autoload :Installer,  "bundler/plugin/installer"
+    autoload :SourceList, "bundler/plugin/source_list"
+
+    class << self
```
Comment (segiddins) [inline lib/bundler/plugin.rb]: since this class can't be instantiated, maybe it should be a module?

File: lib/bundler/plugin/dsl.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module Bundler
+  # Dsl to parse the Gemfile looking for plugins to install
+  class Plugin::Dsl < Bundler::Dsl
+    alias_method :_gem, :gem # To use for plugin installation as gem
+
+    # So that we don't have to override all there methods to dummy ones
+    # explicitly.
+    # They will be handled by missing_methods
```
Comment (segiddins) [inline lib/bundler/plugin/dsl.rb]: `method_missing`

File: lib/bundler/plugin/dsl.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    [:gemspec, :gem, :path, :install_if, :platforms, :env].each {|m| undef_method m }
+
+    def initialize
+      @sources = Plugin::SourceList.new
+
+      super
```
Comment (segiddins) [inline lib/bundler/plugin/dsl.rb]: `super` should be called first

File: lib/bundler/source.rb [hunk 6]
```diff
@@ -7,14 +7,22 @@ class Source
 
     attr_accessor :dependency_names
 
+    def initialize
+      @options = {}
```
Comment (segiddins) [inline lib/bundler/source.rb]: this is a bit of a code smell to me

**Comment by segiddins** [comment]
If at all possible, it'd be great to get rid of all of the `if @options[:plugin]` -- in my eyes, it looks like that's introducing a look of branching where it may not be entirely necessary

**Comment by asutoshpalai** [comment]
`@options[:plugin]` stemmed out of my attempt to use the existing source classes for fetch and install the plugins.

While writing that, I tired real hard to avoid `@options[:plugin]`. But the alternate solutions leads to a lot of code duplication or a major refactor of the the source classes.

I am still in lookout for solution and would definitely appreciate a few pointers.

I expect that development of source plugins might show me some way out of it, but research on those lines up till doesn't seem to thought any light in this direction.

**Comment by asutoshpalai** [comment]
@segiddins : fixed the issues you pointed out (including the code smell :smile_cat: ). Removed `@options[:plugin]`.

The command function is also ready.

File: lib/bundler/cli/install.rb [hunk 7]
```diff
@@ -77,25 +49,15 @@ def run
...
-      Bundler.settings.with       = options[:with]
+      normalize_settings
+
       Bundler::Fetcher.disable_endpoint = options["full-index"]
-      Bundler.settings[:disable_shared_gems] = Bundler.settings[:path] ? true : nil
 
       # rubygems plugins sometimes hook into the gem install process
       Gem.load_env_plugins if Gem.respond_to?(:load_env_plugins)
 
+      Plugin.eval_gemfile(Bundler.default_gemfile) if ENV["BUNDLE_PLUGIN"]
```
Comment (segiddins) [inline lib/bundler/cli/install.rb]: it looks like this is being called for side effects, which is confusing

File: lib/bundler/cli/install.rb [hunk 8]
```diff
@@ -190,5 +154,64 @@ def check_for_options_conflicts
...
+
+      options[:with]    = with
+      options[:without] = without
+    end
+
+    def normalize_settings
+      Bundler.settings[:path]     = nil if options[:system]
+      Bundler.settings[:path]     = "vendor/bundle" if options[:deployment]
+      Bundler.settings[:path]     = options["path"] if options["path"]
+      Bundler.settings[:path] ||= "bundle" if options["standalone"]
```
Comment (segiddins) [inline lib/bundler/cli/install.rb]: the alignment in this method is now off
also, it'd be nice to have a blank line between the different settings?

File: lib/bundler/cli/plugin.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      "URL of the RubyGems source to fetch the plugin from"
+    method_option "version", :type => :string, :default => nil, :banner =>
+      "The version of the plugin to fetch"
+    method_option "git", :type => :string, :default => nil, :banner =>
+      "URL of the git repo to fetch from"
+    method_option "branch", :type => :string, :default => nil, :banner =>
+      "The git branch to checkout"
+    method_option "ref", :type => :string, :default => nil, :banner =>
+      "The git revision to check out"
+    def install(plugin)
```
Comment (segiddins) [inline lib/bundler/cli/plugin.rb]: any reason this couldn't install multiple plugins at once?

File: lib/bundler/plugin.rb [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+module Bundler
+  module Plugin
+    autoload :Api,        "bundler/plugin/api"
+    autoload :Dsl,        "bundler/plugin/dsl"
+    autoload :Index,      "bundler/plugin/index"
+    autoload :Installer,  "bundler/plugin/installer"
+    autoload :SourceList, "bundler/plugin/source_list"
+
+    PLUGIN_FILE_NAME = "plugin.rb".freeze
```
Comment (segiddins) [inline lib/bundler/plugin.rb]: `plugins.rb`?

File: lib/bundler/plugin.rb [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    @commands = {}
+
+    class << self
```
Comment (segiddins) [inline lib/bundler/plugin.rb]: since this is never instantiated, maybe it should be a module with `module_function` instead?

File: lib/bundler/plugin/api.rb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  # previous behavior or if otherwise desired, we can reimplement(or implement)
+  # the method to preserve compatibility.
+  #
+  # To use this, either the class can inherit this class or use it directly.
+  # For example of both types of use, refer the file `spec/plugins/command.rb`
+  #
+  # To use it without inheriting, you will have to create an object of this
+  # to use the functions (except for declaration functions like command, source,
+  # and hooks).
+  class Plugin::Api
```
Comment (segiddins) [inline lib/bundler/plugin/api.rb]: `API`

File: lib/bundler/plugin/dsl.rb [hunk 13]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+module Bundler
+  # Dsl to parse the Gemfile looking for plugins to install
+  class Plugin::Dsl < Bundler::Dsl
```
Comment (segiddins) [inline lib/bundler/plugin/dsl.rb]: `DSL`

File: lib/bundler/plugin/dsl.rb [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+      super
+      @sources = Plugin::SourceList.new
+    end
+
+    def plugin(name, *args)
+      _gem(name, *args)
+    end
+
+    def method_missing(name, *args)
+      # Dummy evaluation
```
Comment (segiddins) [inline lib/bundler/plugin/dsl.rb]: can we please only ignore methods defined on `Bundler::Dsl` ? otherwise this could lead to bugs in this class

File: lib/bundler/plugin/index.rb [hunk 15]
```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+require "yaml"
+
+module Bundler
+  # Manages which plugins are installed and their sources. This also is supposed to map
+  # which plugin does what (currently the features are not implemented so this class is
+  # now a stub class).
+  class Plugin::Index
```
Comment (segiddins) [inline lib/bundler/plugin/index.rb]: please nest the nested class

File: lib/bundler/plugin/index.rb [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    # the functions of the plugin to existing maps and also the name to source location.
+    #
+    # @param [String] name of the plugin to be registered
+    # @param [String] path where the plugin is installed
+    # @param [Array<String>] commands that are handled by the plugin
+    def register_plugin(name, path, commands)
+      @plugin_paths[name] = path
+
+      common = commands & @commands.keys
+      raise "Command(s) #{common.join(", ")} are already registered" if common.any?
```
Comment (segiddins) [inline lib/bundler/plugin/index.rb]: this should raise a specific error

File: lib/bundler/plugin/index.rb [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    end
+
+  private
+
+    # Reads the index file from the directory and initializes the instance variables.
+    def load_index
+      SharedHelpers.filesystem_access(index_file, :read) do |index_f|
+        valid_file = index_f && index_f.exist? && !index_f.size.zero?
+        break unless valid_file
+        index = YAML.load_file(index_f)
```
Comment (segiddins) [inline lib/bundler/plugin/index.rb]: if this can _ever_ be used in the same process as `bundle exec`, we can't use YAML

File: lib/bundler/plugin/installer.rb [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+    # gems to their paths
+    #
+    # @param specs to install
+    #
+    # @return [Hash] map of names to path where the plugin was installed
+    def install_from_spec(specs)
+      paths = {}
+
+      specs.each do |spec|
+        next if spec.name == "bundler"
```
Comment (segiddins) [inline lib/bundler/plugin/installer.rb]: is this really necessary?

File: lib/bundler/cli/install.rb (hunk 7 above)
Comment (asutoshpalai) [inline lib/bundler/cli/install.rb]: I was not able to find any better place for it.

Can you please suggest some

File: lib/bundler/plugin.rb (hunk 11 above)
Comment (asutoshpalai) [inline lib/bundler/plugin.rb]: `module_function` doesn't provide the ability to define private methods on the module. That leaves us with two options: 1. Give up on singleton pattern for Plugin, 2. Make the private methods public.

I am aware that [bbatsov/ruby-style-guide](https://github.com/bbatsov/ruby-style-guide#module-function) recomends it, but I don't find it reflected in [GitHub's](https://github.com/airbnb/ruby) or [Airbnb's](https://github.com/styleguide/ruby)

File: lib/bundler/plugin/installer.rb (hunk 18 above)
Comment (asutoshpalai) [inline lib/bundler/plugin/installer.rb]: I can't remember why I wrote this, maybe some reminiscent of some other installation flow I was trying out.

Anyways it is not necessary now, so removing it

File: lib/bundler/cli/plugin.rb (hunk 9 above)
Comment (asutoshpalai) [inline lib/bundler/cli/plugin.rb]: Only that, I was not able to come up with some flow to accept the version/ref for multiple plugins.

Comment (segiddins) [inline lib/bundler/cli/plugin.rb]: use the same for all of them?

File: lib/bundler/cli/install.rb (hunk 7 above)
Comment (segiddins) [inline lib/bundler/cli/install.rb]: well `eval_gemfile` suggests it will return something, if it doesn't then that method should be renamed

File: lib/bundler/plugin/index.rb [hunk 19]
```diff
@@ -23,7 +30,7 @@ def register_plugin(name, path, commands)
         @plugin_paths[name] = path
 
         common = commands & @commands.keys
-        raise "Command(s) #{common.join(", ")} are already registered" if common.any?
+        raise CommandConflict.new(name, common) if common.any?
```
Comment (segiddins) [inline lib/bundler/plugin/index.rb]: `unless common.empty?`

File: lib/bundler/cli/plugin.rb (hunk 9 above)
Comment (asutoshpalai) [inline lib/bundler/cli/plugin.rb]: or use the default and raise error when multiple plugins are specified and a version is given?

Comment (segiddins) [inline lib/bundler/cli/plugin.rb]: nah since you might want multiple gems at the same version, like if they're all in a suite of gems

File: lib/bundler/cli.rb [hunk 20]
```diff
@@ -82,6 +82,10 @@ def help(cli = nil)
     end
 
     def self.handle_no_command_error(command, has_namespace = $thor_runner)
+      if ENV["BUNDLE_PLUGIN"] && Bundler::Plugin.command?(command)
```
Comment (segiddins) [inline lib/bundler/cli.rb]: could this please using `Bundler.settings`?

File: lib/bundler/cli.rb [hunk 21]
```diff
@@ -437,6 +441,12 @@ def env
       Env.new.write($stdout)
     end
 
+    if ENV["BUNDLE_PLUGIN"]
```
Comment (segiddins) [inline lib/bundler/cli.rb]: settings

File: lib/bundler/plugin/api.rb [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+      def cache
+        Plugin.cache.join("plugins")
+      end
+
+      # A tmp dir to be used by plugins
+      #
+      # @param [String] name unique for the plugin or the purpose
+      # @return [Pathname] object for the new directory created
+      def tmp(name)
+        Bundler.tmp(File.join(["plugin", name]))
```
Comment (segiddins) [inline lib/bundler/plugin/api.rb]: I think this can just be `Bundler.tmp("plugins", name)`

File: lib/bundler/plugin/index.rb [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+          "commands" => @commands,
+        }
+
+        SharedHelpers.filesystem_access(index_file) do |index_f|
+          FileUtils.mkdir_p(index_f.dirname)
+          File.open(index_f, "w") {|f| f.puts dump_yaml(index) }
+        end
+      end
+
+      def dump_yaml(hash)
```
Comment (segiddins) [inline lib/bundler/plugin/index.rb]: the YAML stuff should be refactored so it can be shared with settings, and unit tested

File: lib/bundler/plugin/installer.rb [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+      def install(names, options)
+        version = options[:version] || [">= 0"]
+
+        if options[:git]
+          install_git(names, version, options)
+        elsif options[:source]
+          source = options[:source]
+          install_rubygems(names, version, source)
+        else
+          raise(ArgumentError, "You need to provide the source")
```
Comment (segiddins) [inline lib/bundler/plugin/installer.rb]: shouldn't it default to `Gem.sources` if no source is explicitly provided?

**Comment by segiddins** [comment]
This is looking really good! Just for fun, could you try adding a spec for using plugins in an inline gemfile?

File: lib/bundler/cli.rb (hunk 21 above)
Comment (asutoshpalai) [inline lib/bundler/cli.rb]: I did that in  4a1baea, but some specs were failing because when called here (during the loading of `Cli` class)  `Bundler.settings` gets initialized before `CLI` class gets initialized.

So some options like `--gemfile` don't get properly set. Initializing `Settings` calls `Bundler.root` which caches the root path determined without respecting `--gemfile` option (which is set in [cli#initialize](https://github.com/bundler/bundler/blob/master/lib/bundler/cli.rb#L21)).

So reverted it cf284c0. I tried digging but failed to fix the problem.

File: lib/bundler/plugin/api.rb (hunk 22 above)
Comment (asutoshpalai) [inline lib/bundler/plugin/api.rb]: `Bundler.tmp` doesn't accept splatted arguments. But :+1: on making name splatted, we can use `join`

File: lib/bundler/plugin/index.rb (hunk 23 above)
Comment (asutoshpalai) [inline lib/bundler/plugin/index.rb]: I had already refactored it, just hadn't pushed it yet. I was trying to make the specs pass with `Settings` using it.
Made the specs pass :grinning:

File: lib/bundler/cli.rb (hunk 21 above)
Comment (asutoshpalai) [inline lib/bundler/cli.rb]: We can allow `plugin` sub command and show message if called but not enabled.

**Comment by asutoshpalai** [comment]
Added thee default gem source, spec for inline gemfile and shifted settings to the new serializer. 

The only thing that remains is use of `ENV` as commented above. I believe that we should allow the sub command and show message when used.

**Comment by asutoshpalai** [comment]
Using `Bundler.settings` now :dart:

**Comment by segiddins** [comment]
Cool! Can you just make sure `plugins` is now a bool key in Settings?

File: lib/bundler/plugin.rb [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+    # @param [Hash] options various parameters as described in description
+    # @option options [String] :source rubygems source to fetch the plugin gem from
+    # @option options [String] :version (optional) the version of the plugin to install
+    def install(names, options)
+      paths = Installer.new.install(names, options)
+
+      save_plugins paths
+    rescue PluginError => e
+      paths.values.map {|path| Bundler.rm_rf(path) } if paths
+      Bundler.ui.error "Failed to install plugin #{name}: #{e.message}\n  #{e.backtrace.join("\n  ")}"
```
Comment (segiddins) [inline lib/bundler/plugin.rb]: here only the first  line of the backtrace will be indented

File: lib/bundler/plugin.rb [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+    # @param [Pathname] gemfile path
+    def gemfile_install(gemfile = nil, &inline)
+      if block_given?
+        builder = DSL.new
+        builder.instance_eval(&inline)
+        definition = builder.to_definition(nil, true)
+      else
+        definition = DSL.evaluate(gemfile, nil, {})
+      end
+      return unless definition.dependencies.any?
```
Comment (segiddins) [inline lib/bundler/plugin.rb]: `return if definition.dependencies.empty?`

File: lib/bundler/plugin.rb [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+
+    # Checks if any plugins handles the command
+    def command?(command)
+      !index.command_plugin(command).nil?
+    end
+
+    # To be called from Cli class to pass the command and argument to
+    # approriate plugin class
+    def exec_command(command, args)
+      raise UndefinedCommandError, "Command #{command} not found" unless command? command
```
Comment (segiddins) [inline lib/bundler/plugin.rb]: the interpolated command in backticks

File: lib/bundler/plugin.rb [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+
+    # Checks if the gem is good to be a plugin
+    #
+    # At present it only checks whether it contains plugins.rb file
+    #
+    # @param [Pathname] plugin_path the path plugin is installed at
+    # @raise [Error] if plugins.rb file is not found
+    def validate_plugin!(plugin_path)
+      plugin_file = plugin_path.join(PLUGIN_FILE_NAME)
+      raise MalformattedPlugin, "#{PLUGIN_FILE_NAME} was not found in the plugin!" unless plugin_file.file?
```
Comment (segiddins) [inline lib/bundler/plugin.rb]: no `!`

File: lib/bundler/plugin/api.rb [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+
+      # A tmp dir to be used by plugins
+      # Accepts names that get concatenated as suffix
+      #
+      # @return [Pathname] object for the new directory created
+      def tmp(*names)
+        Bundler.tmp(["plugin", *names].join("-"))
+      end
+
+      def method_missing(name, *args, &blk)
```
Comment (segiddins) [inline lib/bundler/plugin/api.rb]: should also override `respond_to_missing?`

**Comment by segiddins** [comment]
This gets a 👍 from me to ship in 1.13 (but disabled by default). There are some scenarios I'd like to see tested better (plugins with dependencies for example), but that can be done in a later PR. Well done, @asutoshpalai! @fotanus, I'm going to merge this by EOD unless you have any objections?

**Comment by fotanus** [comment]
I haven't check the code closely enough, but I'm with you just because you are the most able person to do such decisions! 

The design is quite good by what I saw in the tests, and deviate little from the original proposal, so I think we are on point.

**Comment by segiddins** [comment]
@homu r+
